src/python/pants/backend/python/tasks2/partition_targets.py

```diff
@@ -0,0 +1,146 @@
+# coding=utf-8
+# Copyright 2016 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import (absolute_import, division, generators, nested_scopes, print_function,
+                        unicode_literals, with_statement)
+
+from collections import namedtuple
+
+from pants.build_graph.target import Target
+from pants.task.task import Task
+
+
+class TargetsPartition(object):
+  """Represents a grouping of targets into non-empty subsets."""
+  def __init__(self, subsets):
+    """Constructs a new TargetsPartition from a given collection of subsets.
+
+    :param subsets: an iterable of iterables of targets.
+    """
+    if any(len(subset) == 0 for subset in subsets):
+      raise ValueError('Subsets must be non-empty.')
+    self._subsets = frozenset(frozenset(subset) for subset in subsets)
+    if self._subsets and (
+        len(frozenset.union(*self._subsets)) != sum(len(subset) for subset in self._subsets)):
+      raise ValueError('Subsets must be disjoint.')
+    self._subset_by_target = {target: subset for subset in self._subsets for target in subset}
+
+  @property
+  def subsets(self):
+    return self._subsets
+
+  def find_subset_for_target(self, target):
+    """Returns a subset containing the given targets.
+
+    Raises a KeyError if such target does not exist."""
+    return self._subset_by_target[target]
+
+  def find_subset_for_targets(self, targets):
+    """Returns a single subset that contains all given targets.
+
+    If such subset does not exists within the partition a ValueError is raised."""
+    if not self._subsets:
+      raise ValueError('No subsets.')
+    if not targets:
+      # Any subset will do
+      return next(iter(self._subsets))
+    subset = self._subset_by_target.get(next(iter(targets)))
+    if not subset or not subset.issuperset(set(targets)):
+      raise ValueError('No subset contains all targets')
+    return subset
+
+
+class PartitionTargets(Task):
+  """Split the target roots into partitions.
+
+  This task implements multiple strategies for partitioning the target roots. A partition is a
+  grouping of the target roots into non-empty disjoint subsets. Interpreter and external
+  requirements are resolved for each subset, and therefore the targets within each subset
+  must be compatible with each other.
+
+  Supported strategies:
+
+  STRATEGY_MINIMAL:
```

Comment:
Yup, same problem today. The algorithms aren't at fault, just the application of them in these situations. I do think running one target root at a time should be the default. We can potentially utilize parallel subprocess runs of roots in follow ups to help speed that up too in advance of the more general ongoing V2 engine work that will make that sort of parallelization free and ~default.